File: app/src/main/java/com/babestudios/companyinfouk/ui/officers/OfficersActivity.java
package com.babestudios.companyinfouk.ui.officers;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.babestudios.companyinfouk.CompaniesHouseApplication;
import com.babestudios.companyinfouk.R;
import com.babestudios.companyinfouk.data.model.officers.OfficerItem;
import com.babestudios.companyinfouk.data.model.officers.Officers;
import com.babestudios.companyinfouk.ui.officerdetails.OfficerDetailsActivity;
import com.babestudios.companyinfouk.uiplugins.BaseActivityPlugin;
import com.babestudios.companyinfouk.utils.DividerItemDecoration;
import com.babestudios.companyinfouk.utils.EndlessRecyclerViewScrollListener;
import com.google.gson.Gson;
import com.pascalwelsch.compositeandroid.activity.CompositeActivity;

import net.grandcentrix.thirtyinch.plugin.TiActivityPlugin;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;

public class OfficersActivity extends CompositeActivity implements OfficersActivityView, OfficersAdapter.OfficersRecyclerViewClickListener {

	@Bind(R.id.toolbar)
	Toolbar toolbar;

	@Bind(R.id.officers_recycler_view)
	RecyclerView officersRecyclerView;

	private OfficersAdapter officersAdapter;

	@Bind(R.id.progressbar)
	ProgressBar progressbar;

	@Inject
	OfficersPresenter officersPresenter;

	String companyNumber;

	TiActivityPlugin<OfficersPresenter, OfficersActivityView> officersActivityPlugin = new TiActivityPlugin<>(
			() -> {
				CompaniesHouseApplication.getInstance().getApplicationComponent().inject(this);
				return officersPresenter;
			});

	BaseActivityPlugin baseActivityPlugin = new BaseActivityPlugin();

	public OfficersActivity() {
		addPlugin(officersActivityPlugin);
		addPlugin(baseActivityPlugin);
	}

	@SuppressWarnings("ConstantConditions")
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_officers);
		baseActivityPlugin.logScreenView(this.getLocalClassName());

		ButterKnife.bind(this);
		if (toolbar != null) {
			setSupportActionBar(toolbar);
			getSupportActionBar().setDisplayHomeAsUpEnabled(true);
			getSupportActionBar().setTitle(R.string.officers);
			toolbar.setNavigationOnClickListener(v -> onBackPressed());
		}
		companyNumber = getIntent().getStringExtra("companyNumber");
		createOfficersRecyclerView();


	}
	private void createOfficersRecyclerView() {
		LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
		officersRecyclerView.setLayoutManager(linearLayoutManager);
		officersRecyclerView.addItemDecoration(
				new DividerItemDecoration(this));

		officersRecyclerView.addOnScrollListener(new EndlessRecyclerViewScrollListener(linearLayoutManager) {
			@Override
			public void onLoadMore(int page, int totalItemsCount) {
				officersActivityPlugin.getPresenter().loadMoreOfficers(page);
			}
		});
	}

	@Override
	public void showProgress() {
		progressbar.setVisibility(View.VISIBLE);
	}

	@Override
	public void hideProgress() {
		progressbar.setVisibility(View.GONE);
	}


	@Override
	public void showOfficers(Officers officers) {
		if (officersRecyclerView.getAdapter() == null) {
			officersAdapter = new OfficersAdapter(OfficersActivity.this, officers);
			officersRecyclerView.setAdapter(officersAdapter);
		} else {
			officersAdapter.updateItems(officers);
		}
	}

	@Override
	public void officersItemClicked(View v, int position, OfficerItem officerItem) {
		Gson gson = new Gson();
		String jsonItem = gson.toJson(officerItem, OfficerItem.class);
		Intent intent = new Intent(this, OfficerDetailsActivity.class);
		intent.putExtra("officerItem", jsonItem);
		baseActivityPlugin.startActivityWithRightSlide(intent);
	}

	@Override
	public String getCompanyNumber() {
		return companyNumber;
	}

	@Override
	public void showError() {
		Toast.makeText(this, R.string.could_not_retrieve_officer_info, Toast.LENGTH_LONG).show();
	}

	@Override
	public void super_onBackPressed() {
		super.super_finish();
		super_overridePendingTransition(R.anim.left_slide_in, R.anim.left_slide_out);
	}
}

File: app/src/main/java/com/babestudios/companyinfouk/ui/officers/OfficersActivity.kt
package com.babestudios.companyinfouk.ui.officers

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.babestudios.base.mvp.ErrorType
import com.babestudios.base.mvp.list.BaseViewHolder
import com.babestudios.base.view.DividerItemDecoration
import com.babestudios.base.view.EndlessRecyclerViewScrollListener
import com.babestudios.base.view.MultiStateView.*
import com.babestudios.companyinfo.core.injection.CoreInjectHelper
import com.babestudios.companyinfouk.R
import com.babestudios.companyinfouk.ext.logScreenView
import com.babestudios.companyinfouk.ext.startActivityWithRightSlide
import com.babestudios.companyinfouk.ui.officerdetails.createOfficerDetailsIntent
import com.babestudios.companyinfouk.ui.officers.list.*
import com.trello.rxlifecycle2.components.support.RxAppCompatActivity
import com.uber.autodispose.AutoDispose
import com.uber.autodispose.ScopeProvider
import com.ubercab.autodispose.rxlifecycle.RxLifecycleInterop
import io.reactivex.CompletableSource
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.activity_officers.*
import kotlinx.android.synthetic.main.multi_state_view_error.view.*

private const val COMPANY_NUMBER = "com.babestudios.companyinfouk.ui.company_number"

class OfficersActivity : RxAppCompatActivity(), ScopeProvider {


	private var officersAdapter: OfficersAdapter? = null

	override fun requestScope(): CompletableSource = RxLifecycleInterop.from(this).requestScope()

	private val viewModel by lazy { ViewModelProviders.of(this).get(OfficersViewModel::class.java) }

	private lateinit var officersPresenter: OfficersPresenterContract

	private val eventDisposables: CompositeDisposable = CompositeDisposable()

	private lateinit var comp: OfficersComponent

	//region life cycle

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_officers)
		logScreenView(this.localClassName)
		setSupportActionBar(pabOfficers.getToolbar())
		supportActionBar?.setDisplayHomeAsUpEnabled(true)
		pabOfficers.setNavigationOnClickListener { onBackPressed() }
		supportActionBar?.setTitle(R.string.officers)
		when {
			viewModel.state.value.officerItems != null -> {
				initPresenter(viewModel)
			}
			savedInstanceState != null -> {
				savedInstanceState.getParcelable<OfficersState>("STATE")?.let {
					with(viewModel.state.value) {
						officerItems = it.officerItems
						companyNumber = it.companyNumber
					}
				}
				initPresenter(viewModel)
			}
			else -> {
				viewModel.state.value.companyNumber = intent.getStringExtra(COMPANY_NUMBER)
				initPresenter(viewModel)
			}
		}

		createRecyclerView()
		observeState()
	}

	override fun onResume() {
		super.onResume()
		observeActions()
	}

	override fun onSaveInstanceState(outState: Bundle) {
		outState.putParcelable("STATE", viewModel.state.value)
		super.onSaveInstanceState(outState)
	}

	private fun initPresenter(viewModel: OfficersViewModel) {
		if (!::officersPresenter.isInitialized) {
			comp = DaggerOfficersComponent
					.builder()
					.coreComponent(CoreInjectHelper.provideCoreComponent(applicationContext))
					.build()
			officersPresenter = comp.officersPresenter()
			officersPresenter.setViewModel(viewModel, requestScope())
		}
	}

	private fun createRecyclerView() {
		val linearLayoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
		rvOfficers?.layoutManager = linearLayoutManager
		rvOfficers.addItemDecoration(DividerItemDecoration(this))
		rvOfficers.addOnScrollListener(object : EndlessRecyclerViewScrollListener(linearLayoutManager) {
			override fun onLoadMore(page: Int, totalItemsCount: Int) {
				officersPresenter.loadMoreOfficers(page)
			}
		})
	}

	override fun onBackPressed() {
		super.finish()
		overridePendingTransition(R.anim.left_slide_in, R.anim.left_slide_out)
	}

	//endregion

	//region render

	private fun observeState() {
		viewModel.state
				.`as`(AutoDispose.autoDisposable(this))
				.subscribe { render(it) }
	}

	private fun render(state: OfficersState) {
		when {
			state.isLoading -> msvOfficers.viewState = VIEW_STATE_LOADING
			state.errorType != ErrorType.NONE -> {
				msvOfficers.viewState = VIEW_STATE_ERROR
				state.errorType = ErrorType.NONE
				msvOfficers.tvMsvError.text = state.errorMessage
			}
			state.officerItems == null -> msvOfficers.viewState = VIEW_STATE_EMPTY
			else -> {
				state.officerItems?.let {
					msvOfficers.viewState = VIEW_STATE_CONTENT
					if (rvOfficers?.adapter == null) {
						officersAdapter = OfficersAdapter(it, OfficersTypeFactory())
						rvOfficers?.adapter = officersAdapter
					} else {
						officersAdapter?.updateItems(it)
					}
					observeActions()
				}
			}
		}
	}

	//endregion

	//region events

	private fun observeActions() {
		eventDisposables.clear()
		officersAdapter?.getViewClickedObservable()
				?.take(1)
				?.`as`(AutoDispose.autoDisposable(this))
				?.subscribe { view: BaseViewHolder<AbstractOfficersVisitable> ->
					viewModel.state.value.officerItems?.let { officerItems ->
						startActivityWithRightSlide(
								this.createOfficerDetailsIntent(
										(officerItems[(view as OfficersViewHolder).adapterPosition] as OfficersVisitable).officersItem))
					}
				}
				?.let { eventDisposables.add(it) }
	}

	//endregion
}

fun Context.createOfficersIntent(companyNumber: String): Intent {
	return Intent(this, OfficersActivity::class.java)
			.putExtra(COMPANY_NUMBER, companyNumber)
}
